adapters/lora/src/main/java/org/eclipse/hono/adapter/lora/LoraCommandSubscriptions.java

```diff
@@ -0,0 +1,180 @@
+                        closeConsumersForTenant(notification.getTenantId());
+                    }
+                });
+
+        NotificationEventBusSupport.registerConsumer(vertx, DeviceChangeNotification.TYPE,
+                notification -> {
+                    if (LifecycleChange.DELETE.equals(notification.getChange())) {
+                        closeConsumersForGateway(notification.getTenantId(), notification.getDeviceId());
+                    }
+                });
+    }
+
+    private void closeConsumersForGateway(final String tenantId, final String gatewayId) {
+        final SubscriptionKey key = new SubscriptionKey(tenantId, gatewayId);
+        final var subscription = commandSubscriptions.remove(key);
+        if (subscription == null) {
+            return;
+        }
+
+        final Context currentCtx = vertx.getOrCreateContext();
+        final Span span = TracingHelper
+                .buildSpan(tracer, null, "close command subscription for deleted gateway",
+                        getClass().getSimpleName())
+                .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_CLIENT)
+                .start();
+        TracingHelper.setDeviceTags(span, tenantId, gatewayId);
+        final Context subscriptionContext = subscription.three();
+        subscriptionContext.runOnContext(x -> {
+            final ProtocolAdapterCommandConsumer commandConsumer = subscription.one();
+            commandConsumer.close(false, span.context()).onComplete(r -> {
+                currentCtx.runOnContext(s -> span.finish());
+            });
+        });
+    }
+
+    @SuppressWarnings("rawtypes")
+    private void closeConsumersForTenant(final String tenantId) {
+        if (!commandSubscriptions.entrySet().stream().anyMatch(e -> e.getKey().getTenant().equals(tenantId))) {
+            return;
+        }
+
+        final Context currentCtx = vertx.getOrCreateContext();
+        final Span span = TracingHelper
+                .buildSpan(tracer, null, "close command subscriptions for disabled/deleted tenant",
+                        getClass().getSimpleName())
+                .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_CLIENT)
+                .start();
+        TracingHelper.setDeviceTags(span, tenantId, null);
+        final List<Future> consumerCloseFutures = Collections.synchronizedList(new ArrayList<>());
+        final var iter = commandSubscriptions.entrySet().iterator();
+        while (iter.hasNext()) {
+            final var subscription = iter.next();
+            if (!subscription.getKey().getTenant().equals(tenantId)) {
+                continue;
+            }
+
+            final Context subscriptionContext = subscription.getValue().three();
+            subscriptionContext.runOnContext(x -> {
+                final ProtocolAdapterCommandConsumer commandConsumer = subscription.getValue().one();
+                consumerCloseFutures.add(commandConsumer.close(false, span.context()));
```

Comment:
Adding an entry to `consumerCloseFutures` here in the `subscriptionContext` thread means that in the `CompositeFuture.join(consumerCloseFutures)` code below, the `consumerCloseFutures` list might still by empty.

`CompositeFuture.join(emptyList)` is just a succeeded Future, therefore the `onComplete` action would be run here before the commandConsumer got closed.

Therefore, the `consumerCloseFutures` list has to be filled outside of the `subscriptionContext.runOnContext` block  (i.e. on the same thread as `CompositeFuture.join(consumerCloseFutures)`).
You can do this by adding a Promise future
```
            final Promise<Void> closePromise = Promise.promise();
            consumerCloseFutures.add(closePromise.future());
```
and completing the `closePromise` with the `commandConsumer.close()` result.